File: benchmarks/cross-language/go/ping_pong.go
package main

import (
	"fmt"
	"sync"
	"time"
)

const MESSAGES = 10000000

func main() {
	fmt.Println("=== Go Ping-Pong Benchmark ===")
	fmt.Printf("Messages: %d\n", MESSAGES)
	fmt.Println("Using goroutines with buffered channels")
	fmt.Println()

	chanA := make(chan int, 1)
	chanB := make(chan int, 1)

	var wg sync.WaitGroup
	wg.Add(2)

	start := time.Now()

	go func() {
		defer wg.Done()
		for i := 0; i < MESSAGES; i++ {
			chanA <- i
			<-chanB
		}
	}()

	go func() {
		defer wg.Done()
		for i := 0; i < MESSAGES; i++ {
			<-chanA
			chanB <- i
		}
	}()

	wg.Wait()
	elapsed := time.Since(start)

	totalNs := float64(elapsed.Nanoseconds())
	nsPerMsg := totalNs / MESSAGES
	throughput := 1e9 / nsPerMsg
	cyclesPerMsg := nsPerMsg * 3.0

	fmt.Printf("Cycles/msg:     %.2f\n", cyclesPerMsg)
	fmt.Printf("Throughput:     %.2f M msg/sec\n", throughput/1e6)
}

File: benchmarks/cross-language/elixir/ping_pong.exs
#!/usr/bin/env elixir

# Elixir ping-pong benchmark using processes
defmodule PingPong do
  @messages 10_000_000

  def ping(pong_pid, 0) do
    send(pong_pid, :done)
  end

  def ping(pong_pid, n) do
    send(pong_pid, {:ping, self()})
    receive do
      :pong -> ping(pong_pid, n - 1)
    end
  end

  def pong(ping_pid) do
    receive do
      {:ping, sender} ->
        send(sender, :pong)
        pong(ping_pid)
      :done ->
        :ok
    end
  end

  def rdtsc() do
    # Use monotonic time in nanoseconds
    :erlang.monotonic_time(:nanosecond)
  end

  def run() do
    IO.puts("=== Elixir Ping-Pong Benchmark ===")
    IO.puts("Messages: #{@messages}")
    IO.puts("Using Erlang/OTP processes\n")

    ping_pid = self()
    pong_pid = spawn(fn -> pong(ping_pid) end)

    start = rdtsc()
    ping(pong_pid, @messages)
    finish = rdtsc()

    total_ns = finish - start
    ns_per_msg = total_ns / @messages
    throughput = 1.0e9 / ns_per_msg
    cycles_per_msg = ns_per_msg * 3.0  # Approximate at 3GHz

    IO.puts("Cycles/msg:     #{Float.round(cycles_per_msg, 2)}")
    IO.puts("Throughput:     #{Float.round(throughput / 1.0e6, 2)} M msg/sec")
  end
end

PingPong.run()
